### DISCUSSION TITLE
Confused by certificates

### DISCUSSION THREAD

**Discussion by B-Down** [jlesage/docker-firefox#207] [Q&A] [answered]
Hello,

I've had a couple FF containers running for quite some time now, and all is fine... except I've recently been wanting to get some secured connection to those, and that's when my brain started exploding **><ins>.</ins><**

I duplicated one of my tried-and-true containers and set its `SECURE_CONNECTION` env variable to `1`; for starters.
Then I read your readme and went on to get ssvnc running.
That's where things got... complicated.

In your readme, you mention that `/config/certs/vnc-server.pem` is the be-all, end-all when it comes to certificates.
Even though my `/config/certs/` is different (I have `vnc-fullchain.pem` and `vnc-privkey.pem` there), but that could be because I'm using an old image.
Anyway, I wouldn't know what to do with those...

Now, on ssvnc's side, creating a new self-signed certificate seems like a good idea.
So I did, and copied the public `.pem` to `/config/certs` in my container (with no clue as to whether it actually did *something*).

I managed to get my ssvnc to accept my container's certificate, but it still can't actually connect to it: I get a `ReadExact: Socket Error While Reading` and that's it.
Prior to that, I was trying to connect to `Docker_IP:5912`, because I can't read (5912 is my container's external listening port), but at least I was getting `Connections: accepted` in my Docker logs, even though there was also a `Clean disconnection` in the next second...

Now (connecting to `Docker_IP:12`), all I get is `Connections: blacklisted`, and no clue on how to clear that (or even correct it, if need be).

Here's an excerpt (`tail`) of said log:
> `[xvnc        ] Tue Dec 12 17:30:32 2023`
> `[xvnc        ]  Connections: accepted: /tmp/vnc.sock`
> `[xvnc        ] Tue Dec 12 17:30:33 2023`
> `[xvnc        ]  VNCSConnST:  closing /tmp/vnc.sock: Clean disconnection`
> `[xvnc        ]  EncodeManager: Framebuffer updates: 0`
> `[xvnc        ]  EncodeManager:   Total: 0 rects, 0 pixels`
> `[xvnc        ]  EncodeManager:          0 B (1:-nan ratio)`
> `[xvnc        ]  Connections: closed: /tmp/vnc.sock`
> `[xvnc        ]  ComparingUpdateTracker: 0 pixels in / 0 pixels out`
> `[xvnc        ]  ComparingUpdateTracker: (1:-nan ratio)`
> `[xvnc        ] Tue Dec 12 18:26:12 2023`
> `[xvnc        ]  Connections: accepted: /tmp/vnc.sock`
> `[xvnc        ]  VNCSConnST:  closing /tmp/vnc.sock: Clean disconnection`
> `[xvnc        ]  EncodeManager: Framebuffer updates: 0`
> `[xvnc        ]  EncodeManager:   Total: 0 rects, 0 pixels`
> `[xvnc        ]  EncodeManager:          0 B (1:-nan ratio)`
> `[xvnc        ]  Connections: closed: /tmp/vnc.sock`
> `[xvnc        ]  ComparingUpdateTracker: 0 pixels in / 0 pixels out`
> `[xvnc        ]  ComparingUpdateTracker: (1:-nan ratio)`
> `[xvnc        ] Tue Dec 12 18:29:58 2023`
> `[xvnc        ]  Connections: accepted: /tmp/vnc.sock`
> `[xvnc        ] Tue Dec 12 18:29:59 2023`
> `[xvnc        ]  VNCSConnST:  closing /tmp/vnc.sock: Clean disconnection`
> `[xvnc        ]  EncodeManager: Framebuffer updates: 0`
> `[xvnc        ]  EncodeManager:   Total: 0 rects, 0 pixels`
> `[xvnc        ]  EncodeManager:          0 B (1:-nan ratio)`
> `[xvnc        ]  Connections: closed: /tmp/vnc.sock`
> `[xvnc        ]  ComparingUpdateTracker: 0 pixels in / 0 pixels out`
> `[xvnc        ]  ComparingUpdateTracker: (1:-nan ratio)`
> `[xvnc        ] Tue Dec 12 18:32:24 2023`
> `[xvnc        ]  Connections: blacklisted: /tmp/vnc.sock`
> `[xvnc        ] Tue Dec 12 18:43:06 2023`
> `[xvnc        ]  Connections: accepted: /tmp/vnc.sock`
> `[xvnc        ]  VNCSConnST:  closing /tmp/vnc.sock: Clean disconnection`
> `[xvnc        ]  EncodeManager: Framebuffer updates: 0`
> `[xvnc        ]  EncodeManager:   Total: 0 rects, 0 pixels`
> `[xvnc        ]  EncodeManager:          0 B (1:-nan ratio)`
> `[xvnc        ]  Connections: closed: /tmp/vnc.sock`
> `[xvnc        ]  ComparingUpdateTracker: 0 pixels in / 0 pixels out`
> `[xvnc        ]  ComparingUpdateTracker: (1:-nan ratio)`
> `[xvnc        ] Tue Dec 12 18:44:49 2023`
> `[xvnc        ]  Connections: blacklisted: /tmp/vnc.sock`
> `[xvnc        ] Tue Dec 12 18:46:14 2023`
> `[xvnc        ]  Connections: blacklisted: /tmp/vnc.sock`
> `[xvnc        ] Tue Dec 12 18:46:37 2023`
> `[xvnc        ]  Connections: blacklisted: /tmp/vnc.sock`
 
TL;DR: I'm at a loss as to how to actually configure ssvnc and/or my container(s) to connect to the latter securely.

**Comment by jlesage** [comment]
Note that if you don't provide any certificates, the container will automatically generate (self-signed) ones.  So I would first use this method to test and make sure you are able to connect.  You can remove any certificate from `/config` and restart the container.

> but it still can't actually connect to it: I get a `ReadExact: Socket Error While Reading` and that's it.

Are you using the Windows version? If yes, you need to use the version mentioned here: https://github.com/jlesage/docker-firefox#ssvnc.  Otherwise, you get the error message you are seeing.

**Comment by B-Down** [comment]
Thanks for the feedback.

> Are you using the Windows version?

I've been using your version.

> So I would first use this method to test and make sure you are able to connect.

That's one of my initial woes... I don't know what to do with those auto-generated certificates.

I can get a `.crt` file when I click on "Fetch Cert" in ssvnc, and I set it as **ServerCert**, but whatever I try to put in **MyCert** (be it `vnc-fullchain.pem`, `vnc-privkey.com`, or even the ones *__I__* created), it doesn't go through (`Failed to connect to server (127.0.0.1:32)`).

For instance, when using the "public" cert from the container:
> `[ ] Running on Windows 6.1`
> `[ ] No limit detected for the number of clients`
> `[.] stunnel 5.49 on x86-pc-msvc-1500 platform`
> `[.] Compiled/running with OpenSSL 1.0.2p-fips  14 Aug 2018`
> `[.] Threading:WIN32 Sockets:SELECT,IPv6 TLS:ENGINE,FIPS,OCSP,PSK,SNI`
> `[ ] errno: (*_errno())`
> `[ ] GUI message loop initialized`
> `[ ] Running on Windows 6.1`
> `[.] Reading configuration from file stunnel-vnc-32.conf`
> `[.] UTF-8 byte order mark not detected`
> `[.] FIPS mode disabled`
> `[ ] Compression disabled`
> `[ ] No PRNG seeding was required`
> `[ ] Initializing service [vnc32]`
> `[ ] Ciphers: HIGH:!aNULL:!SSLv2:!DH:!kDHEPSK`
> `[ ] TLS options: 0x83000BFF (+0x80000BFF, -0x00000000)`
> `[ ] Loading certificate from file: F:/Files/Portable/ssvnc/Windows/certs/firefox-1-ssl-cert-pub.pem`
> `[ ] Certificate loaded from file: F:/Files/Portable/ssvnc/Windows/certs/firefox-1-ssl-cert-pub.pem`
> `[ ] Loading private key from file: F:/Files/Portable/ssvnc/Windows/certs/firefox-1-ssl-cert-pub.pem`
> `[!] error queue: 140B0009: error:140B0009:SSL routines:SSL_CTX_use_PrivateKey_file:PEM lib`
> `[!] SSL_CTX_use_PrivateKey_file: 906D06C: error:0906D06C:PEM routines:PEM_read_bio:no start line`
> `[!] Service [vnc32]: Failed to initialize TLS context`
> `[ ] Deallocating section defaults`
> 
> `[!] Server is down`

And the private one:
> (...)
> `[ ] Loading certificate from file: F:/Files/Portable/ssvnc/Windows/certs/firefox-1-ssl-cert.pem`
> `[!] error queue: 140DC009: error:140DC009:SSL routines:SSL_CTX_use_certificate_chain_file:PEM lib`
> `[!] SSL_CTX_use_certificate_chain_file: 906D06C: error:0906D06C:PEM routines:PEM_read_bio:no start line`
> (...)

**Comment by jlesage** [comment, marked as answer]
Just tried on my side: In SSVNC, the only thing I did is:

-  Make sure `Use SSL` is selected
-  Unchecked `Verify All Certs`

It then successfully connected to the container.

**Comment by B-Down** [comment]
> Unchecked Verify All Certs

Damn, that did the trick.
I can't say I understand what the rationale is behind this (we're using certs, but <ins>not</ins> *verifying* them ?!), but as long as it works...

I'll run some more tests before marking this as answered, but thanks for (inadvertently ?) pointing that tick out, in any case.

**Reply by jlesage** [reply]
The next step would be to save the certificate.  In SSVNC, click `Fetch Cert`, then click the `Save` button.

Once the certificate is saved, you can keep `Verify All Certs` checked.

I also verified this on my side and it worked.  I kept all default certificate folders.

**Comment by B-Down** [comment]
> In SSVNC, click Fetch Cert, then click the Save button.

I believe I did that early on, but I'll test the whole process again.


Also, we're most likely not in *certificate territory* anymore, but for some reason after a time (usually between 10 and 20 minutes), the connection just closes itself, while my non-SSL plain ol' TigerVNC session stays on no matter what.

> `2023.12.13 12:54:29 LOG6[0]: TLS closed (SSL_read)`
> `2023.12.13 12:54:29 LOG6[0]: Read socket closed (readsocket)`
> `2023.12.13 12:54:29 LOG6[0]: SSL_shutdown successfully sent close_notify alert`
> `2023.12.13 12:54:29 LOG5[0]: Connection closed: 9456 byte(s) sent to TLS, 1309206 byte(s) sent to socket`

Do you happen to suffer from that as well ?
I couldn't find a way to have some sort of definitive keepalive on the stunnel "users" mailing list...

Another issue I have is that the connection is choppy at best (typically: scrolling a page), and setting the compression levels to rock-bottom doesn't improve that one bit.
I understand SSL implies more trafic, but that's on a static page, and my network isn't quite 10Base2...


Back to certificates, I've tried to use both my own and the ones from the container, in TigerVNC Viewer (using [this guide](https://github.com/TigerVNC/tigervnc/wiki/Secure-your-connection)), to no avail.
That could be because TigerVNC Server works (much ?) differently than xvnc on the container, for all I know (yes, I did add my own certificates to the container's `/config/cert/`).
Perhaps stunnel can be used alongside TigerVNC the same way it works with ssvnc, but that may be well outside the scope of my "skills".

**Comment by B-Down** [comment]
I managed to get both *UltraVNC Viewer* and *TigerVNC Viewer* to work by setting them in `Options -> Advanced -> Change VNC Viewer` in ssvnc and now it's real smooth (compared to the *TightVNC Viewer*, at least).
On its password frame TigerVNC keeps telling me the connection is not secured, but I assume it can only see its own way of connecting to the server, and is oblivious to the _stunnel_ it's going through.

Next step: the test of time... gotta see if it can stay around without getting (seemingly) randomly murdered.

**Comment by jlesage** [comment]
TigerVNC viewer is probably easier to use.  Add the `SECURE_CONNECTION_VNC_METHOD=TLS` environment variable to the container and you should be able to connect with TigerVNC (with default settings) without any problem.

**Comment by B-Down** [comment]
That parameter sure helped, thanks for that.
Prior to using it I had the "usual" disconnections, while UltraVNC was stable from the get-go.
But since I'm more used to TigerVNC, it's even better.

In any case, thanks a lot for your time and knowledge helping the noob *me* on those matters !

**Comment by B-Down** [comment]
@jlesage Out of mere curiosity, would you happen to know if there's a way to make TigerVNC "silent" ?
When I close the secured session (through ssvnc), I always get an error frame... which looks an awful lot more like a status report than an error (and of course there's no such thing on regular, non-tunneled connections):
![2023-12-18 09 11 59](https://github.com/jlesage/docker-firefox/assets/128323366/6e8d1d67-cad6-45f6-94c2-7e0e54b7eaa1)
While it's already annoying as it is, it doesn't close stunnel either, since it's waiting for that frame to go away...

On a slightly different note, would there be an *easy* way to remove the IP(s) blacklisted from/by the container when I'm performing tests and such ?
I'm hoping for something more practical than restarting the container.
